### PR TITLE
Refactor the Swap example to swap arbitrary ERC-20 and swap from EVM to Bitcoin

### PR DESCRIPTION
**Pull request by fadeev**
* Switched back to providing a target ZRC-20 address instead of chain ID, because a user might want to swap to a specific ZRC-20 (like USDC) on a destination chain
* Used `npx hardhat omnichain Swap targetToken:address recipient:address` to leverage templating
* Remove `MinAmountOut` as it doesn't contribute to the example and can be easily added (we've been using 0 as a value anyways).
* Right now you need to provide a contract address of the target ZRC-20, because we need a way to convert human readable token name to a ZRC-20 address.
* Added support to swap from EVM to Bitcoin

Related docs: https://github.com/zeta-chain/docs/pull/189

### CODE REVIEW

**Comment by fadeev** [comment]
Testing BTC: 1c15ef97b09f1c6c8f922a791c33d6a4ae8be9bc77f6965582379d1f5b8bcca1

**Comment by fadeev** [comment]
Testing BTC:

https://blockstream.info/testnet/tx/12c778438891f7f1166294234d3b74e6c5c16f111aeaa16e17381dfb203dcfbb

**Comment by fadeev** [comment]
BTC swap seems to be working:

```
npx hardhat cctx 7dc7942a1e6c4d1532dadd1ec1c2cac1a7302ef04d4845a15935c9147a48f9b4
✓ CCTXs on ZetaChain found.

✓ 0x370f41bf277f8905d59e8ea9ebf1006d4609d4681ac0a735b21e23f0e3943971: 18332 → 7001: OutboundMined (Remote omnichain contract call completed)
✓ 0x804eb819dda20ed6dbe003ee2cb0bb8d9a97cf452bc62f82ed3e8c4355075fe7: 7001 → 5: OutboundMined 
```

**Review by andresaiello — APPROVED** [review]

**Comment by fadeev** [comment]
I've made a few tweaks to support swapping from EVM chains to Bitcoin by changing the expected second value in the `message` to `bytes`. Now if you want to swap EVM -> EVM you send a hex address, which will be stored as `bytes`. If you swap from EVM -> you send a long bech32 address encoded as ASCII bytes, it is also stored as `bytes` in the contract.

## BTC to gETH

```
npx hardhat send-btc --amount 0.001 --recipient tb1qy9pqmk2pd9sv63g27jt8r657wy0d9ueeh0nqur --memo cC02751bAA435E9A5cF3bd22F96a21d7C002E15013A0c5930C028511Dc02665E7285134B6d11A5f42cD3D070aE1BD365909dD859d29F387AA96911e1

npx hardhat cctx b52b9025fd0707f9922b2f52f7dd7a944454dedcd6248443bb7bb8652cde7017
✓ CCTXs on ZetaChain found.

✓ 0x5e5f157d7255f97d13e646f9e1809bb27df82cb33855f394d9798231b44c0a82: 18332 → 7001: OutboundMined (Remote omnichain contract call completed)
✓ 0x186b582201d3abbae0d5f19aa8f2569cd579296fd731183bf4d409013cb160b3: 7001 → 5: PendingOutbound  → OutboundMined 
```

## gETH to tMATIC

```
npx hardhat interact --contract 0xcC02751bAA435E9A5cF3bd22F96a21d7C002E150 --amount 0.1 --target-token 0x48f80608B672DC30DC7e3dbBd0343c5F02C738Eb --recipient 0x2cD3D070aE1BD365909dD859d29F387AA96911e1 --network goerli_testnet

npx hardhat cctx 0x7ebd2bff64cbc530c145a60e4830ba2ddc536bc62cf8c5566c900143b0e08baf
✓ CCTXs on ZetaChain found.

✓ 0x5082897440218490193a724a22b7ed3f8744760956d857199e76bf2453f901b2: 5 → 7001: OutboundMined (Remote omnichain contract call completed)
✓ 0x3aebbba04cd284d2e87b9c8414f6946bdf933efad45076aba7c691e5a32895ba: 7001 → 80001: PendingOutbound  → OutboundMined 

## gETH to BTC

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inTxHashToCctx/0xe247091341c92308652b96506309a1a9066e7d69eec810a92b0717a6cda45009

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/cctx/0xac248959e2858f557956539451f4d105e45002ece23e1c19c3dcd2d8ea88e13d

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/cctx/0x1243a5ee33d8b57c0bce5e69c16a45b984a3fbf846c0f6f9fdb7ecf42e60e20f

(CCTX tracking failed for this one, need to debug the toolkit, but this is not related)

File: scripts/integration.sh [hunk 1]
```diff
@@ -13,148 +13,147 @@ npx hardhat compile --force
...
-OMNI_SWAP_TX=$(npx hardhat interact --contract $OMNI_SWAP_CONTRACT --network goerli_testnet --amount 0.01 --recipient $SENDER --destination mumbai_testnet --json | jq -r '.hash')
+OMNI_SWAP_TX=$(npx hardhat interact --contract $OMNI_SWAP_CONTRACT --network goerli_testnet --amount 0.01 --recipient $SENDER --target-token 0x48f80608B672DC30DC7e3dbBd0343c5F02C738Eb --json | jq -r '.hash')
 echo $OMNI_SWAP_TX
 OMNI_SWAP_CCTX=$(npx hardhat cctx $OMNI_SWAP_TX --json)
 echo $OMNI_SWAP_CCTX
 
 popd
 
-pushd "./omnichain/staking"
+# pushd "./omnichain/staking"
```
Comment (andresaiello) [inline scripts/integration.sh:23]: it's intentional all the lines commented?

Comment (fadeev) [reply scripts/integration.sh]: This was an accidental change, I've reverted https://github.com/zeta-chain/example-contracts/pull/85/commits/b145cbf94f07e1e7b6e07334c88b5ba930af014b Thanks for catching it!

**Comment by fadeev** [comment]
@andresaiello please, review.

**Review by andresaiello — APPROVED** [review]

**Comment by magj2006** [comment]
can we support to swap if specified source token address and destination token address ? and the amount represents amount of source token if specified.

**Comment by fadeev** [comment]
@magj2006 the source token address is passed in the `zrc20` param of the `onCrossChainCall` function:

https://github.com/zeta-chain/example-contracts/blob/b145cbf94f07e1e7b6e07334c88b5ba930af014b/omnichain/swap/contracts/Swap.sol#L29

Is there a reason why we would need to pass it explicitly?

**Comment by magj2006** [comment]
> @magj2006 the source token address is passed in the `zrc20` param of the `onCrossChainCall` function:
> 
> https://github.com/zeta-chain/example-contracts/blob/b145cbf94f07e1e7b6e07334c88b5ba930af014b/omnichain/swap/contracts/Swap.sol#L27-L32
> 
> Is there a reason why we would need to pass it explicitly?

If I want to bridge USDC on goerli to USDC on bsc_testnet, how I pass parameters?  As I understand when we use command `npx hardhat interact --contract 0xcC02751bAA435E9A5cF3bd22F96a21d7C002E150 --amount 0.1 --target-token 0x07865c6e87b9f70255377e024ace6630c1eaa37f --recipient 0x2cD3D070aE1BD365909dD859d29F387AA96911e1 --network goerli_testnet`, source token is gETH, which is gasCoin of goerli and amount argument indicates the number of gETH, not of UDSC.

**Comment by fadeev** [comment]
@magj2006 that's a good point! This, however, only requires changes to the `interact` task, not the contract itself. The reason I didn't do it yet is mostly because I want to decide on how to tell the task which token you want to swap. I think we can do something like this:

```
npx hardhat interact --contract 0xcC02751bAA435E9A5cF3bd22F96a21d7C002E150 --amount 0.1 --from usdc --to geth --recipient 0x2cD3D070aE1BD365909dD859d29F387AA96911e1 --network goerli_testnet
```

**Comment by fadeev** [comment]
Created an issue for this: https://github.com/zeta-chain/example-contracts/issues/97

**Comment by magj2006** [comment]
> @magj2006 that's a good point! This, however, only requires changes to the `interact` task, not the contract itself. The reason I didn't do it yet is mostly because I want to decide on how to tell the task which token you want to swap. I think we can do something like this:
> 
> ```
> npx hardhat interact --contract 0xcC02751bAA435E9A5cF3bd22F96a21d7C002E150 --amount 0.1 --from usdc --to geth --recipient 0x2cD3D070aE1BD365909dD859d29F387AA96911e1 --network goerli_testnet
> ```

You are right. the hard point is how to calculate gas with USDC for this situation.

**Pull request merged**